### DISCUSSION TITLE
How to exclude actions from triggering draft changes in field layouts?

### DISCUSSION THREAD

**Discussion by BenParizek** [craftcms/cms#11349] [Q&A] [answered]
I have a custom Element that uses Drafts where I want to give the user the option to take an action from the Element edit page. Currently, any change on that page triggers a draft resave and presents the user with the banner at the top of the edit page which says: _Showing your unsaved changes. Discard._

For context, my element is an Email Element and I'm trying to give the user the option to send a test email to preview the content they are working on. The problem currently is that if they click on the 'Test' button, it pops up a modal to allow them to do that but also triggers the _Showing your unsaved changes_ forcing them to need to take an additional action to discard the changes after they finish sending the test even though they didn't change anything related to their content.

I think this is the relevant section of code:

https://github.com/craftcms/cms/blob/6db32e613a15f496e91fb39fe53899dfe6f2d683/src/web/assets/cp/src/js/ElementEditor.js#L252-L268

And it appears the status icons are excluded from triggering these types of changes:

```
if ($(ev.target).is(this.statusIcons())) {
  return;
}
```

Is there any way that I could signal to the Craft JS that I do not want an element on the Element Edit page to trigger a 'things have changed' signal?

One thing that could be nice is if I could just add some class or data attribute that Craft checks for and excludes similar to the status icons. Something like: 

```<div ignore-element-editor-attribute>My thing that triggers a non-content related action</div>```

**Comment by brandonkelly** [comment, marked as answer]
It will only trigger an autosave if the form data has changed. What exactly is changing in your form data, and could that data be added to the DOM outside of the primary form instead?

**Comment by BenParizek** [comment]
I'm currently inserting a button and a modal form using the Element::getAdditionalButtons() method which is inside of the primary form tag.

I've just tested having the button trigger an html form defined in my javascript and that appears to get around this issue. I'll explore that direction more. 

Are there any ways within the Element APIs I can insert data outside of the primary form or is javascript going to be the primary way to achieve that?

**Reply by brandonkelly** [reply]
You’ll just need to add it with some JS code.

**Comment by michaelrog** [comment]
It might be possible to wrap your inputs content in a "sink" — i.e. a `div` that absorbs DOM events and [stops them from propagating](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation) further up the chain. Either way... gonna need some custom JS.
